include/up-cpp/communication/RpcClient.h

```diff
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache License Version 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_CLIENT_RPCCLIENT_H
+#define UP_CPP_CLIENT_RPCCLIENT_H
+
+#include <up-cpp/datamodel/builder/Payload.h>
+#include <up-cpp/datamodel/builder/UMessage.h>
+#include <up-cpp/transport/UTransport.h>
+#include <up-cpp/utils/Expected.h>
+#include <uprotocol/v1/umessage.pb.h>
+#include <uprotocol/v1/ustatus.pb.h>
+
+#include <future>
+#include <memory>
+#include <string>
+#include <variant>
+
+namespace uprotocol::communication {
+
+/// @brief Interface for uEntities to invoke RPC methods.
+///
+/// Like all L2 client APIs, the RpcClient is a wrapper on top of the L1
+/// UTransport API; in this instance, it is the request-initiating half of the
+/// RPC model.
+struct RpcClient {
+	/// @brief Constructs a client connected to a given transport
+	///
+	/// @param method UUri of the method this client will invoke.
+	/// @param priority Priority of request messages when invoking methods.
+	/// @param ttl Time from the moment `invokeMethod()` is called that the
+	///            request will still be considered valid. Must be >0.
+	/// @param payload_format (Optional) If sending a payload, this sets the
+	///                       format that will be expected when invokeMethod()
+	///                       is called. Empty request payloads can only be
+	///                       sent if this was not set.
+	/// @param permission_level Permission level of this client
+	/// @param token TAP token for accessing restricted services
+	///
+	/// For guidance on the permeission_level and token parameters, see:
+	/// https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/permissions.adoc
+	explicit RpcClient(std::shared_ptr<transport::UTransport> transport,
+	                   v1::UUri&& method, v1::UPriority priority,
+	                   std::chrono::milliseconds ttl,
+	                   std::optional<v1::UPayloadFormat> payload_format = {},
+	                   std::optional<uint32_t> permission_level = {},
+	                   std::optional<std::string> token = {});
+
+	/// @brief as found in v1::UAttributes
+	using Commstatus = uint32_t;
+
+	/// @brief Contains either a UMessage (when successful) or a UStatus /
+	/// Commstatus when an error occurred.
+	using StatusOrMessage =
+	    utils::Expected<v1::UMessage, std::variant<v1::UStatus, Commstatus>>;
```

Comment:
`UStatus` contains a `UCode`, right?

Can we have this look like:
```suggestion
	    utils::Expected<v1::UMessage, v1::UStatus>;
```
instead?

I am just thinking this could be a bit of a cognitive burden on having to check the `variant`. WDYT?

Reply:
commstatus is not defined as a UCode. It's just an arbitrary uint32

Reply:
:neutral_face: 

It changed.

Reply:
TBH I'm not sure why commstatus isn't just a whole UStatus. The non-enum fields are all optional, so it wouldn't significantly increase message sizes.

In any case, I'm going to go change that now :confused:

Reply:
I'm actually still going to keep the types a little separate so it is possible to tell the difference between a status returned locally and a status received in a message. But it'll be a v1::UStatus and a v1::UCode now.

Reply:
> :neutral_face: 
> 
> It changed.

Caught me off guard too when I went looking 😅